src/internals/symbols.ts

```diff
@@ -0,0 +1,26 @@
+/**
+ * Symbols used internally within ts-pattern to construct and discriminate
+ * Guard, Not, and Select, and AnonymousSelect patterns
+ *
+ * Symbols have the advantage of not appearing in auto-complete suggestions in
+ * user defined patterns, and eliminate the risk of property
+ * overlap between ts-pattern internals and user defined patterns.
+ *
+ * These symbols have to be visible to tsc for type inference to work, but
+ * users should not import them
+ * @module
+ * @private
+ * @internal
+ */
+
+export const toExclude = Symbol('@ts-pattern/to-exclude');
+export type toExclude = typeof toExclude;
+
+export const matcher = Symbol('@ts-pattern/matcher');
```

Comment:
To ease interop, would it be possible to make this one `Symbol.matcher ?? Symbol('@ts-pattern/matcher')`?

Reply:
This release doesn't include custom matcher protocol support on purpose, and this symbol isn't supposed to be use externally for now, hence the name of the `internals/` folder ;)

Reply:
I agree it would be good to make ts-pattern support a kind of custom matcher protocol that would be optionally implemented by othre libraries, without needing a hard dependency on ts-pattern. A "type only" dependency could be also work.

Even though ts-pattern v4 is architected with this goal in mind, a matcher protocol isn't publicly exposed in the intial version and I don't want to tie the release with the support of this protocol, which imo is much harder to get right and to iterate on.

Reply:
I'm guessing should the tc39 proposal make progress to stage 3/4 (its up for stage 2 at the end of this month) that there might still be space for libraries that can create or expose various objects that implement the matcher protocol.